### PR TITLE
Split WorldQuery into WorldQueryData and WorldQueryFilter

### LINKED ISSUE #7680: Use distinct types for query "data" generics and query "filter" generics
## What problem does this solve or what need does it fill?

It's easy for users to accidentally mix up their "data" (`Q`) and "filter" (`F`) generics on `Query`.

For example:

- `Query<(&A, With<B>)`: forces ugly unpacking.
- `Query<(&A, Without<B>)`: likely seriously wrong.
- `Query<Changed<T>)`: probably not what the user wants.
- `Query<(), Entity>`:  nonsensical
- `Query<&A, &B>`: very weird, probably broken

See #7679 for an example of this footgun in the wild.

## What solution would you like?

1. Add `WorldQueryData: WorldQuery` and `WorldQueryFilter: WorldQuery`.
2. Split `derive(WorldQuery)` macro into two parts.
3. Only implement `WorldQueryData` / `WorldQueryFilter` for the correct `WorldQuery` types.

## What alternative(s) have you considered?

We could use a custom linter #1602, but a) this catches errors later / less elegantly and b) we don't have one yet.

### PR DESCRIPTION
**Pull request by wainwrightmark**
# Objective

- Fixes #7680

## Solution

The traits  `WorldQueryData : WorldQuery` and `WorldQueryFilter : WorldQuery` have been added and some of the types and functions from `WorldQuery` has been moved into them. 

`ReadOnlyWorldQuery` has been replaced with `ReadOnlyWorldQueryData`. 

`WorldQueryFilter` is safe (as long as `WorldQuery` is implemented safely). 

`WorldQueryData` is unsafe - safely implementing it requires that `Self::ReadOnly` is a readonly version of `Self` (this used to be a safety requirement of `WorldQuery`)

The type parameters `Q` and `F` of `Query` must now implement `WorldQueryData` and `WorldQueryFilter` respectively.

This makes it impossible to accidentally use a filter in the data position or vice versa which was something that could lead to bugs. ~~Compile failure tests have been added to check this.~~

It was previously sometimes useful to use `Option<With<T>>` in the data position. Use `Has<T>` instead in these cases.

The derive macro has been split into separate derive macros for `WorldQueryData` and `WorldQueryFilter`.

Previously it was possible to derive both `WorldQuery` for a struct that had a mixture of data and filter items. This would not work correctly in some cases but could be a useful pattern in others. *This is no longer possible.*

---

## Notes

- The changes outside of `bevy_ecs` are all changing type parameters to the new types, updating the macro use, or replacing `Option<With<T>>` with `Has<T>`.

- All `WorldQueryData` types always returned `true` for `IS_ARCHETYPAL` so I moved it to `WorldQueryFilter` and 
replaced all calls to it with `true`. That should be the only logic change outside of the macro generation code. 

- `Changed<T>` and `Added<T>` were being generated by a macro that I have expanded. Happy to revert that if desired. 

- The two derive macros share some functions for implementing `WorldQuery` but the tidiest way I could find to implement them was to give them a ton of arguments and ask clippy to ignore that. 

### Changed
-  Split `WorldQuery` into `WorldQueryData` and `WorldQueryFilter` which now have separate derive macros. It is not possible to derive both for the same type.
- `Query` now requires that the first type argument implements `WorldQueryData` and the second implements  `WorldQueryFilter`

## Migration Guide

- Update derives

```rust
// old
#[derive(WorldQuery)]
#[world_query(mutable, derive(Debug))]
struct CustomQuery {
    entity: Entity,
    a: &'static mut ComponentA
}

#[derive(WorldQuery)]
struct QueryFilter {
    _c: With<ComponentC>
}

// new 
#[derive(WorldQueryData)]
#[world_query_data(mutable, derive(Debug))]
struct CustomQuery {
    entity: Entity,
    a: &'static mut ComponentA,
}

#[derive(WorldQueryFilter)]
struct QueryFilter {
    _c: With<ComponentC>
}
```
- Replace `Option<With<T>>` with `Has<T>`

```rust
/// old
fn my_system(query: Query<(Entity, Option<With<ComponentA>>)>)
{
  for (entity, has_a_option) in query.iter(){
    let has_a:bool = has_a_option.is_some();
    //todo!()
  }
}

/// new
fn my_system(query: Query<(Entity, Has<ComponentA>)>)
{
  for (entity, has_a) in query.iter(){
    //todo!()
  }
}
```

- Fix queries which had filters in the data position or vice versa.

```rust
// old
fn my_system(query: Query<(Entity, With<ComponentA>)>)
{
  for (entity, _) in query.iter(){
  //todo!()
  }
}

// new
fn my_system(query: Query<Entity, With<ComponentA>>)
{
  for entity in query.iter(){
  //todo!()
  }
}

```

### CODE REVIEW

File: crates/bevy_ecs/src/query/fetch.rs [hunk 1]
```diff
@@ -18,18 +17,11 @@ use std::{cell::UnsafeCell, marker::PhantomData};
 ///
 /// - **Component references.**
 ///   Fetches a component by reference (immutably or mutably).
-/// - **`WorldQuery` tuples.**
-///   If every element of a tuple implements `WorldQuery`, then the tuple itself also implements the same trait.
-///   This enables a single `Query` to access multiple components and filter over multiple conditions.
+/// - **`WorldQueryData` tuples.**
+///   If every element of a tuple implements `WorldQueryData`, then the tuple itself also implements the same trait.
+///   This enables a single `Query` to access multiple components.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:22]:
```suggestion
///   This enables a single `Query` to access multiple components.
///   Note that only entities with **all** of the provided components will be fetched.
```

File: crates/bevy_ecs/src/query/fetch.rs [hunk 2]
```diff
@@ -695,26 +516,19 @@ pub struct RefFetch<'w, T> {
     this_run: Tick,
 }
 
-/// SAFETY: `Self` is the same as `Self::ReadOnly`
+/// SAFETY: Updates component access correctly
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:519]: I'd liike something a bit more tangible and helpful here.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 3]
```diff
@@ -859,26 +682,19 @@ pub struct WriteFetch<'w, T> {
     this_run: Tick,
 }
 
-/// SAFETY: access of `&T` is a subset of `&mut T`
+/// SAFETY: Updates component access correctly
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:685]: This safety comment is much less clear than previously.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 4]
```diff
@@ -1168,26 +995,19 @@ unsafe impl<T: ReadOnlyWorldQuery> ReadOnlyWorldQuery for Option<T> {}
 /// ```
 pub struct Has<T>(PhantomData<T>);
 
-// SAFETY: `Self::ReadOnly` is the same as `Self`
+/// SAFETY: Updates component access correctly
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:998]: Another safety comment to improve :)

File: crates/bevy_ecs/src/query/filter.rs [hunk 5]
```diff
@@ -383,274 +466,404 @@ macro_rules! impl_query_filter_tuple {
...
+}
 
-            const IS_DENSE: bool = {
-                match T::Storage::STORAGE_TYPE {
-                    StorageType::Table => true,
-                    StorageType::SparseSet => false,
-                }
-            };
+/// SAFETY: Updates component access correctly
+unsafe impl<T: Component> WorldQuery for Added<T> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/filter.rs:549]: Why isn't this a `WorldQueryFilter` impl?

**Comment by alice-i-cecile** [comment]
Neat, this is impressive work. I like how much clearer (and error-resistant) the end user facing code becomes.

We should consider splitting out the `Has<T>` changes into a seperate PR and merging that ASAP: that's clearly a better way to write this pattern in general.

I was suspicious of the fact that `WorldQuery` itself still hangs around, but after investigating more the shared methods make a lot of sense and I think this is a sensible structure.

**Comment by wainwrightmark** [comment]
Just to explain the safety comments. The old `WorldQuery` has two requirements
```
/// # Safety
///
/// Component access of `Self::ReadOnly` must be a subset of `Self`
/// and `Self::ReadOnly` must match exactly the same archetypes/tables as `Self`
///
/// Implementor must ensure that
/// [`update_component_access`] and [`update_archetype_component_access`]
/// exactly reflects the results of the following methods:
///
/// - [`matches_component_set`]
/// - [`fetch`]
```

The first requirement has moved to `WorkdQueryData`, the second requirement remains on WorldQuery. The old safety comments generally only addressed the first requirement so I've moved them onto the WorldQueryData implementations. There don't seem to be any existing comments about the other safety requirements so I wrote "Updates component access correctly" which is not particularly detailed but I didn't feel super comfortable going into more detail as it's not an area I touched - I just kept the existing implementations.

**Comment by wainwrightmark** [comment]
I spent some time trying to understand the safety requirements of `WorldQuery`. I've codified them and updated all the safety comments on the implementations. 

The general rules for a safe implementation are:

> Implementor must ensure that
> `update_component_access`, `update_archetype_component_access`, `matches_component_set`, and `fetch`
> obey the following:
>
> - For each component mutably accessed by `fetch`, `update_component_access` should add write access unless read or write access has already been added, in which case it should panic.
> - For each component readonly accessed by `fetch`, `update_component_access` should add read access unless write access has already been added, in which case it should panic.
> - For each component mutably accessed by `fetch`, `update_archetype_component_access` should add write access if that component belongs to the archetype.
> - For each component readonly accessed by `fetch`, `update_archetype_component_access` should add read access if that component belongs to the archetype.
> - If `fetch` mutably accesses a component and also accesses the same component again, `update_component_access` should panic.
> - `update_component_access` may not add a `With` filter for a component unless `matches_component_set` always returns `false` if the component set doesn't contain that component.
> - `update_component_access` may not add a `Without` filter for a component unless `matches_component_set` always returns `false` when the component set contains that component.
> - `update_component_access` may replace the filters with a disjunction where every member of the disjunction is a conjunction of the previous filters and the filters of a soundly implemented `WorldQuery` so long as `matches_component_set` is a disjunction of the implementations in those `WorldQuery`s.

And as an example, the safety comment for `&mut T` is:

> `fetch` accesses a single component mutably.
> This is sound because `update_component_access` and `update_archetype_component_access` add write access for that component and panic when appropriate.
> `update_component_access` adds a `With` filter for a component.
> This is sound because `matches_component_set` returns whether the set contains that component.

As far as I can tell these rules are correct and all the existing implementations follow them but I may have missed something. There is also probably scope to make them more succinct (especially the last one).

**Comment by Trashtalk217** [comment]
What is the status of this PR?

**Comment by wainwrightmark** [comment]
> What is the status of this PR?

I think I was fairly happy with this last time I pushed but I imagine there have been some significant changes (e.g. new query types added) since then. I'm happy to rebase and get it back up to date if people want.

**Comment by Trashtalk217** [comment]
Sounds great, I'd be willing to give a review.

**Comment by Trashtalk217** [comment]
I've briefly looked through all the changes and I'll be honest, it's a bit of a mess. It looks like something went wrong with rebasing, since from the 7k code changes, only ~700 lines have anything to do with queries.

Can you clean up this PR? In all honesty, if I were you I'd close this PR, start a new one and lift over the relevant changes. But you may be better at git-fu than I am.

Also, if you don't have time to work on this PR, let me know. I could potentially adopt this PR if you want.

**Comment by wainwrightmark** [comment]
> I've briefly looked through all the changes and I'll be honest, it's a bit of a mess. It looks like something went wrong with rebasing, since from the 7k code changes, only ~700 lines have anything to do with queries.
> 
> Can you clean up this PR? In all honesty, if I were you I'd close this PR, start a new one and lift over the relevant changes. But you may be better at git-fu than I am.
> 
> Also, if you don't have time to work on this PR, let me know. I could potentially adopt this PR if you want.

Yes, something definitely has gone wrong somewhere (possibly I merged instead of rebased at some point?). 
I will do as you suggest and open a new PR.

If you did want to look at the old changes, the page here is much more helpful https://github.com/bevyengine/bevy/compare/main...wainwrightmark:bevy:split-worldquery-into-data-and-filter#diff-b02ae2ab7ec1b2e91fc2f00da4d9483eea1f8f13336c8ec5445b507d608c67a8

**Comment by wainwrightmark** [comment]
Closing in favour of https://github.com/bevyengine/bevy/pull/9918

**Pull request closed without merging**